Sources/Apodini/Request/Request.swift

```diff
@@ -5,28 +5,75 @@
 //  Created by Paul Schmiedmayer on 7/12/20.
 //
 
-import Vapor
+import Foundation
+import NIO
+import Fluent
+@_implementationOnly import Vapor
 
+/// A type representing a server response.
+public protocol Request {
+    /// parameters
+    func parameter<T: Codable>(for parameter: Parameter<T>) throws -> T
+    /// The EventLoop associated with this request
+    var eventLoop: EventLoop { get }
+    /// database
+    var database: Any? { get }
+}
+
+enum RequestDecodingError: Error {
+    case couldNotDecodeParameter(for: UUID)
+}
+
+typealias VaporRequest = Vapor.Request
+
+extension VaporRequest: Request {
+    public func parameter<T: Codable>(for parameter: Parameter<T>) throws -> T {
+        switch parameter.option(for: PropertyOptionKey.http) {
+        case .path where T.self is LosslessStringConvertible:
+            guard let result = parameters.get(parameter.id.uuidString) as? T else {
+                throw RequestDecodingError.couldNotDecodeParameter(for: parameter.id)
+            }
+            return result
+        case .query:
+            guard let name = parameter.name else {
+                throw RequestDecodingError.couldNotDecodeParameter(for: parameter.id)
+            }
+            return try query.get(T.self, at: name)
+        case .body:
+            let length = self.body.data?.readableBytes ?? 0
+            let body = try? self.body.data?.getJSONDecodable(T.self, at: 0, length: length)
+            let string = self.body.string as? T
+            guard let result = body ?? string else {
+                throw RequestDecodingError.couldNotDecodeParameter(for: parameter.id)
+            }
+            return result
+        default:
+            throw RequestDecodingError.couldNotDecodeParameter(for: parameter.id)
+        }
+    }
```

Comment:
Just throwing in some thoughts, haven't had time to look at the rest (updated parts) of the PR yet.

IMO we still need to think about how we would incorporate that into the shared semantic model. Not having a `PropertyOptionKey.http` option defined is actually a common case. The inference approach what type of parameter to use is actually decided and stored in the `EndpointParameter`. Additionally the exporter type decides on where to pull the parameter information from.

As of the changes made today, the shared model will create a request handler, which is responsible to handle, guards, the `handle()` call and response modifiers, thus including the injection of RequestInjectables (and thus Parameters). How exactly those things are handled isn't set in stone.

So we need two abilities here:
* the `.parameter` method needs to be controlled by the shared model. It can match up the uuid with what is stored in its model and then decide on what type of parameter is used.
* The described procedure needs to be able to delegate retrieval of the actual parameter to the interface exporter. I could imagine adding something like `func retrieveParameter<T: Codable>(of endpointType: EndpointParameterType, as type: T, from request: ExporterRequest) -> T` to the `InterfaceExporter` where `ExporterRequest` is what the interface exporter uses as request type e.g. `Vapor.Request`.

Looking at those requirements I think a better approach would be to just simply use a wrapper approach for our own Request type instead of making the request types of the exporters (like `Vapor.Request`) to conform to a protocol. The wrapper approach would allow for easily adding new exporters with different Request types in the future.
But then I don't know what discussions or arguments were brought up in the past that lead to the current approach.

Just wanted to throw as I can't see the current approach working with what we have/need and I don't want to waste your time.

Reference:
* [How the SharedSemanticModelBuilder creates the request handler](https://github.com/Apodini/Apodini/blob/2cc2d6bda74f990bcd87cf177530fa1e4f49f8e2/Sources/Apodini/Semantic%20Model%20Builder/SharedSemanticModelBuilder.swift#L109-L134)
* [Location where a InterfaceExporter would create its own request handler from that](https://github.com/Apodini/Apodini/blob/2cc2d6bda74f990bcd87cf177530fa1e4f49f8e2/Sources/Apodini/Semantic%20Model%20Builder/EndpointsTree.swift#L67-L69)

Reply:
Thanks for the clarification. This is very helpful. I'll try to make the InterfaceExporter to handle parameters and decoding / encoding tomorrow.